### ISSUE TITLE
Iterate through all possible solutions?

### ISSUE THREAD

**Issue by vorner** [luteberget/minisat-rs#1]
Hello

Would it be possible to add an API that iterates through all possible solutions?

Or alternatively, know that a variable is fixed in all solutions or that in some solutions it might be true and in other false (I know I can use the assumptions for that and „try“ it both ways, but that seems inconvenient)?

Thank you

**Comment by luteberget** [comment]
I suppose it is possible to iterate all possible solutions by listing all variables in the problem instance and including a clause containing the negation of all of them. However, when I've had use something like this myself, there was always a distinction between output variables, which characterized the solution, and intermediate variables. So then you would have to post-process to find equal solutions, and it wouldn't be as efficient as adding a custom "reject solution" clause. What use case do you have in mind?

The second part I don't understand. What use case do you have in mind?

**Comment by vorner** [comment]
Recently I've played the alchemists board game. The gist of the game is, the game (a mobile phone app) generates a matching between 8 substances (chicken claw, frog, mushroom, …) and 8 combination of elements (red green and blue can be either + or -). The matching is secret to start with, so all 8! combinations are possible. But as the game progresses, pieces of information about the matching are uncovered and it is up to the players to determine the matching or parts of it (or risk a guess). As a help, a player is given a paper with the substances on top and possible element combinations at the side and can cross out the already impossible pairs.

The problem is, this is error prone and I've also wondered (because of the guess) how probable each matching pair is when there are already some known restrictions. So I'm thinking about writing a helper to keep the track and find out what combinations are still possible and mark the certain pairings and probabilities of the ones that are still possible but uncertain.

I'd create a boolean variable for each substances-elements pair (therefore, there would be 64 of these variables), add the basic rules (exactly one of each line and exactly one of each column). Then I'd keep adding the information as the game would be progressing ‒ setting variables I know to be false to false, adding clauses that these two variables are impossible to be true at the same time, etc. And, after each such addition, I'd like to iterate through all the still possible solutions to see what variables are always true (which I can do by setting it to false and seeing there's *no* solution) and what is the ratio of number of solution with true and false for the probabilities.

Does my explanation make sense?

**Comment by luteberget** [comment]
OK, I think I roughly understand the problem domain -- sounds like a fun challenge. :)  I see how the SAT API is not a perfect match for the problem. When you have few constraints, enumerating solutions will probably also be somewhat inefficient. If you really want to do it this way anyway, it might work out OK, but I don't think the library should encourage it.

Two other suggestions for you:

1. You probably have thought of this, but with a maximum of 8! solutions, brute forcing should be fast enough (maybe faster if you have few constraints) and easier to write. 
2. Binary decision diagrams might be better suited things related to counting solutions and checking whether a variable is constrained, though I'm not very familiar with them. See http://javabdd.sourceforge.net/

**Comment by vorner** [comment]
So, do I understand you correctly that the lower level C API doesn't have this support in some way? My „dream“ API extension would look like something like `fn solve_all(&mut self) -> impl Iterator<Item = Model>` (which could be empty). I think it could be useful for other things too (checking if the solution is the only one or others are possible would need to iterate just first 2 solutions).

The second point I was describing at the original post was something like each variable in the Model knowing if it can „flip“ in some other solution, or if this one is fixed through all of them. Like, a is true in this solution and is true in all solutions, while b is false in this solution, but some other solutions may have it as true.

Again, I don't know what the low-level library supports, so I don't know if anything like this is possible at all. Do you think it makes sense for me to have a look at the low-level API and docs?

I could brute force and I actually considered the SAT library as a crutch for doing so. Not only would it help me to keep and evaluate the constraints as the game progresses conveniently, I imagined it would be able to avoid trying out a lot of the possibilities early on, based on the constraints it already knows. At the beginning, I would get all the 8! solutions, but this would go down fast with the constraints.

**Comment by luteberget** [comment]
No,there is no solve_all function, and conflict-driven clause learning solvers such as MiniSat typically do not directly support enumerating models. You can either add the negation clause containing all variables, or you can try something more intricate like this: https://arxiv.org/abs/1510.00523. Also, MiniSat is very small so you should definitely not hesitate to look at the interface.
